eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Recipient.scala

```diff
@@ -0,0 +1,178 @@
+  }
+}
+
+/** A payment recipient that doesn't expect to receive a payment and can directly be found in the routing graph. */
+case class SpontaneousRecipient(nodeId: PublicKey,
+                                totalAmount: MilliSatoshi,
+                                expiry: CltvExpiry,
+                                preimage: ByteVector32,
+                                customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
+  override val features = Features.empty
+  override val extraEdges = Nil
+
+  private def validateRoute(route: Route): Either[OutgoingPaymentError, Route] = {
+    route.hops.lastOption match {
+      case Some(hop) if hop.nextNodeId == nodeId => Right(route)
+      case Some(hop) => Left(InvalidRouteRecipient(nodeId, hop.nextNodeId))
+      case None => Left(EmptyRoute)
+    }
+  }
+
+  override def buildPayloads(paymentHash: ByteVector32, route: Route): Either[OutgoingPaymentError, PaymentPayloads] = {
+    validateRoute(route).map(_ => {
+      val finalPayload = NodePayload(nodeId, FinalPayload.Standard.createKeySendPayload(route.amount, totalAmount, expiry, preimage, customTlvs))
+      Recipient.buildPayloads(totalAmount, expiry, finalPayload, route.hops)
+    })
+  }
+}
+
+/** A payment recipient that can be reached through a given trampoline node (usually not found in the routing graph). */
+case class ClearTrampolineRecipient(invoice: Bolt11Invoice,
+                                    totalAmount: MilliSatoshi,
+                                    expiry: CltvExpiry,
+                                    trampolineHop: NodeHop,
+                                    trampolinePaymentSecret: ByteVector32,
+                                    customTlvs: Seq[GenericTlv] = Nil) extends Recipient {
+  require(trampolineHop.nextNodeId == invoice.nodeId, "trampoline hop must end at the recipient")
+
+  val trampolineNodeId = trampolineHop.nodeId
+  val trampolineFee = trampolineHop.fee(totalAmount)
+  val trampolineAmount = totalAmount + trampolineFee
+  val trampolineExpiry = expiry + trampolineHop.cltvExpiryDelta
+
+  override val nodeId = invoice.nodeId
+  override val features = invoice.features
+  override val extraEdges = Seq(ExtraEdge(trampolineNodeId, nodeId, ShortChannelId.generateLocalAlias(), trampolineFee, 0, trampolineHop.cltvExpiryDelta, 1 msat, None))
+
+  private def validateRoute(route: Route): Either[OutgoingPaymentError, NodeHop] = {
+    route.finalHop_opt match {
+      case Some(trampolineHop: NodeHop) => Right(trampolineHop)
+      case None => Left(MissingTrampolineHop(trampolineNodeId))
+    }
+  }
+
+  override def buildPayloads(paymentHash: ByteVector32, route: Route): Either[OutgoingPaymentError, PaymentPayloads] = {
+    for {
+      trampolineHop <- validateRoute(route)
+      trampolineOnion <- createTrampolinePacket(paymentHash, trampolineHop)
+    } yield {
```

Comment:
I think that's expected, `for` is a replacement for `flatMap`, not for `map`? Combining `flatMap`s is a bit cumbersome because it adds more layers of indentation, that's why I used `for` where it was necessary.